Initial commit for serverless pattern

diwakarkvs opened the pull request:
*Issue #, if available:* New serverless pattern - apigw-authorizer-sqs

*Description of changes:* This queue levelling pattern can be used when customers want to use APi Gateway RestAPI to send messages to Amazon SQS and have APIGW authorize the request.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

mavi888 commented on cdk-apigw-authorizer-sqs/README.md-OLD:
```diff
@@ -0,0 +1,17 @@
+## My Project
```
please remove this file

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
```diff
@@ -0,0 +1,60 @@
+# AWS Service 1 to AWS Service 2
```
fix title for readme

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
```diff
@@ -0,0 +1,60 @@
+
+This pattern << explain usage >>
+
+Learn more about this pattern at Serverless Land Patterns: << Add the live URL here >>
```
whole readme is missing

mavi888 commented:
@diwakarkvs please fill the README.md file correctly and make sure that there are no extra files commited in ths PR. 

After that is fixed i will continue my review. 

Cheers

diwakarkvs commented:
@mavi888 , I updated the README.md and deleted README copy.md and READEME.md-OLD files. 
I had to add a few pictures to provide context to README content. So i checked in 3 images to /images folder.

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
```diff
@@ -0,0 +1,97 @@
+* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
+* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+* [CDK Installed](https://docs.aws.amazon.com/cdk/v2/guide/getting_started.html) 
+* [Create Lambda authorizer from blueprint](https://github.com/awslabs/aws-apigateway-lambda-authorizer-blueprints/blob/master/blueprints/python/api-gateway-authorizer-python.py)
```
why this cannot be created as IaC?

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,97 @@
+```
+2. Change directory
+```
+cd pattern-apigw-auth-sqs
````
this is not the right folder for this pattern

saikotag replied:
Done! updated the path to cdk-apigw-authorizer-sqs

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,97 @@
+
+1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
+```
+git clone https://github.com/aws-samples/pattern-apigw-auth-sqs.git
````
put the folder serverless-patterns

saikotag replied:
Done! updated the path to https://github.com/aws-samples/serverless-patterns/cdk-apigw-authorizer-sqs.git

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,97 @@
+    cdk synth ApigwSqsStack -o ./cloudformation_templates
+    ```
+4. To deploy AWS resources as a CDK project
+   ##### Paramters needed for CDK 
````
parameters

saikotag replied:
Done! corrected the spelling

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,97 @@
+    ```
+4. To deploy AWS resources as a CDK project
+   ##### Paramters needed for CDK 
+    * storagequeue=Name of the SQS queue where you want data to be sent
````
what queue? is not in the prerequistes. Can you create it as Iac?

saikotag replied:
The SQS queue will be created by IaC. This parameter will take a value for 'Queue Name' - Users can provide a user friendly name that will be created by IaC.

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
```diff
@@ -0,0 +1,97 @@
+4. To deploy AWS resources as a CDK project
+   ##### Paramters needed for CDK 
+    * storagequeue=Name of the SQS queue where you want data to be sent
+    * awsregion=AWS region where you want the APIGW, SQS and token authoizer to be created.
```
can you not get this one from the cdk env?

mavi888 commented on cdk-apigw-authorizer-sqs/apigw_sqs/apigw_sqs_auth_stack.py:
```diff
@@ -0,0 +1,114 @@
+    
+        )       
+        authorizer_lambda_arn = core.CfnParameter(self, "lambdaauthorizerarn", type="String",
+            default='arn:aws:lambda:us-east-1:764551143200:function:apigwRequestAuthorizer',
```
what account name is this one?

saikotag replied:
Deleted the default value

mavi888 commented:
This pattern is not yet ready for merging. 
Please fix the comments and then is the folder cloudformation templates needed? if not please remove.

Also please remove any extra files that is not needed for the project to work or is not a required one listed in here https://github.com/aws-samples/serverless-patterns/tree/main/_pattern-model. Such as contributing, code of conduct, or licence. 

is source.bat required? 

This project has an enourmous amount of files that im not sure they are required for this pattern to work. please clean

saikotag commented:
Hi Mavi,
Updated the code and README.md file based on your comments and inputs. Please review it when you get a chance. Thanks

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,81 @@
+```
+cd cdk-apigw-authorizer-sqs
+```
+3. To generate a cloudformation templates (optional)
````
Can you take the region from the configuration and make CDK create the queue name. You can also use a config file inside the app to pass the parameters as this is not a clean way to do the call. 
Same for the deploying

mavi888 commented on cdk-apigw-authorizer-sqs/README.md:
```diff
@@ -0,0 +1,81 @@
+# CDK Python project for deploying Queue-Based Leveling pattern!
```
remove the exclamation mark

saikotag replied:
Done.

mavi888 commented:
Still some comments to fix in this PR.

Also why the authorizer function cannot be created as IaC in the project? The solution will be more elegant if that is added to it.

saikotag commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,81 @@
+```
+4. To deploy AWS resources as a CDK project
+```
+aws_region_name=<aws_region_name> storage_sqs_queue_name=<user_friendly-SQS_Queue_name> authorizer_lambda_arn=<arn:aws:lambda:<aws_region>:<aws_account>:function:<lambda_function_name> cdk deploy ApigwSqsAuthStack
````
Done! Added instructions to import modules

saikotag commented on cdk-apigw-authorizer-sqs/README.md:
````diff
@@ -0,0 +1,81 @@
+
+Test the api invocation using curl:
+```
+curl -X POST "https://<apigateway-id>.execute-api.us-east-1.amazonaws.com/prod/sqs" -H "Authorization:testtoken"  -H "Content-Type: application/json" --data '{"status": "ok"}'
````
Done.

saikotag commented:
Since APIGW is using custom Lambda authorizer, where users can implement their own token validation logic, I feel that it is best to have users implement their own lambda logic that best suits them. Although, the lambda function can be created as  IaC they still have to customize and write their own code, instead as 'Requirements' we have a step
[Create Lambda authorizer from blueprint](https://github.com/awslabs/aws-apigateway-lambda-authorizer-blueprints/blob/master/blueprints/python/api-gateway-authorizer-python.py)

mavi888 commented:
This pattern was merged and published in this link: https://serverlessland.com/patterns/cdk-apigw-authorizer-sqs

Thanks a lot for contributing!
